pkg/controller/worker/machines_test.go

```diff
@@ -818,6 +861,29 @@ func expectGetSecretCallToWork(c *mockclient.MockClient, awsAccessKeyID, awsSecr
 		})
 }
 
+// expectDeleteAllCallToWork is the intialization methods which mocks the
+// delete call to delete all awsMachineClass from awsMachineClassList
+func expectDeleteAllCallToWork(c *mockclient.MockClient, awsMachineClassList *machinev1alpha1.AWSMachineClassList) {
+	c.EXPECT().
+		DeleteAllOf(context.TODO(), gomock.AssignableToTypeOf(&machinev1alpha1.AWSMachineClass{}), gomock.Any()).
+		AnyTimes().Do(func(_ context.Context, awsMachineClass *machinev1alpha1.AWSMachineClass, namespace client.InNamespace) error {
+		allAreSameKind := true
+		mockObject := machinev1alpha1.AWSMachineClass{}
+		for _, mockObject = range awsMachineClassList.Items {
+			if mockObject.Kind != awsMachineClass.Kind {
+				allAreSameKind = false
+				break
+			}
+		}
+
+		if allAreSameKind {
+			awsMachineClassList.Items = []machinev1alpha1.AWSMachineClass{}
+		}
+
+		return nil
+	})
```

Comment:
I don't understand this logic. In the test you mock `DeleteAllOf ` call to set the awsMachineClassList Items to empty slice and then you expect `Expect(len(mockAWSMachineClassList.Items)).To(Equal(0))` which will be always true as it is the behaviour you defined in the mock of DeleteAllOf. Shouldn't we rather expect DeleteAllOf to be called with the expected args?

Reply:
My main motive for writing this test cases was to make sure that we don't end up calling the wrong `MachineClass` kind from the main ccode. However, yes you are right. I probably am unable to capture my intention with the tests. 

I have tried to improve it with this - https://github.com/gardener/gardener-extension-provider-aws/pull/212/files#diff-12a1fa4ba4f4250731a53989c476c9407e832ccecae17ad3fc1b2eee100ad7c3R929-R932. I have also added some additional (negative) cases to capture failure scenarios.